pkg/controller/iam/managed.go

```diff
@@ -0,0 +1,232 @@
+	}
+	_, err := e.iam.UpdateServiceAccount(ctx, sapb)
+	return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
+}
+
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return errors.New(errNotServiceAccount)
+	}
+
+	deleteRequest := &adminpb.DeleteServiceAccountRequest{
+		Name: cr.Status.AtProvider.Name,
+	}
+
+	err := e.iam.DeleteServiceAccount(ctx, deleteRequest)
+	if status.Code(err) == codes.NotFound {
+		return nil
+	}
+
+	return errors.Wrap(err, errDelete)
+}
+
+// isUpToDate returns true if the supplied Kubernetes resource does not differ
+//  from the supplied GCP resource. It considers only fields that can be
+//  modified in place without deleting and recreating the Service Account.
+func isUpToDate(in *v1beta1.ServiceAccountParameters, observed *adminpb.ServiceAccount) bool {
+	// see comment in serviceaccount_types.go
+	/*
+		if *in.DisplayName != observed.DisplayName {
+			return false
+		}
+		if in.Description != observed.Description {
+			return false
+		}
+		return true
+	*/
+	return *in.DisplayName == observed.DisplayName
+}
+
+// ProviderClient is satisfied by the GCP adminv1.IamClient type or test mocks etc
+type ProviderClient interface {
+	GetServiceAccount(ctx context.Context, req *adminpb.GetServiceAccountRequest, opts ...gax.CallOption) (*adminpb.ServiceAccount, error)
+	CreateServiceAccount(ctx context.Context, req *adminpb.CreateServiceAccountRequest, opts ...gax.CallOption) (*adminpb.ServiceAccount, error)
+	UpdateServiceAccount(ctx context.Context, req *adminpb.ServiceAccount, opts ...gax.CallOption) (*adminpb.ServiceAccount, error)
+	DeleteServiceAccount(ctx context.Context, req *adminpb.DeleteServiceAccountRequest, opts ...gax.CallOption) error
+}
+
+// NewProviderClient returns a new IAM Admin Client (responsible for Service Account management).
+// Credentials must be passed as JSON encoded data.
+func NewProviderClient(ctx context.Context, credentials []byte) (ProviderClient, error) {
+	return adminv1.NewIamClient(ctx, option.WithCredentialsJSON(credentials))
+}
+
+// FullyQualifiedName allows the controller to generate the fully qualified name of the service account
+// before the service account has been created. The fully qualified name has the following format:
+// projects/{project_id}/serviceAccounts/{account}
+// Once a service account has been created, API responses include this value in the `Name` field of the
+// provider response, but before that point (or if a CR is declared to describe a service account that already exists)
+// the controller needs a value to bootstrap the lookup.
+func FullyQualifiedName(cr v1beta1.ServiceAccount, projectID string) string {
+	if cr.Status.AtProvider.Name != "" {
+		return cr.Status.AtProvider.Name
+	}
+	return fmt.Sprintf("projects/%s/serviceAccounts/%s@%s.iam.gserviceaccount.com",
+		projectID, cr.Spec.ForProvider.AccountID, projectID)
```

Comment:
Per comment in `types.go`, I think `cr.Spec.ForProvider.AccountID` will become `meta.GetExternalName(cr)` here

Reply:
`AccountID` is specified by the dev/operator, only used in the `Create()` call, whereas `Name` is the fully qualified string we get back from GCP after the object is created, which we also use to lookup the resource in subsequent observe/update/delete calls. So, I wonder if we should keep `AccountID`, but replace `Name` with the ExternalName annotation (and remove `Name` from the CRD)?

Reply:
check out the new code, I think your feedback on this has been addressed